### ISSUE TITLE
ClickHouse fails on restart if a parameterized view was created

### ISSUE THREAD

**Issue by areshayevcaci** [ClickHouse/ClickHouse#61321]
> Please make sure that the version you're using is still supported (you can find the list [here]
24.2.1.2248 (revision: 54483)

**Describe what's wrong**
Cannot restart clickhouse if parameterized view was created
sample of views:

```
create or replace view v_my_test_view as
with cte_params as 
(select calc_year from my_prm_table
  where calc_year = {p_calc_year:String}
)
select * from my_values_table t
where t.calc_year in (select p.calc_year from cte_params p);

create or replace view v_my_test_view2 as
select * from my_values_table t
where t.calc_year in (select calc_year from my_prm_table p where p.calc_year = {p_calc_year:String});

create or replace view v_my_test_view3 as
select * from my_values_table t
where t.calc_year = {p_calc_year:String}

```
the views works ok until clickhouse restart
when we try to restart database we get error

_<Error> Application: Code: 695. DB::Exception: Load job 'load table dev.v_my_test_view' failed: Code: 49. DB::Exception: Invalid storage definition in metadata file: it's a bug or result of manual intervention in metadata files: Cannot attach table `dev`.`v_my_test_view` from metadata file_

**Does it reproduce on the most recent release?**
24.2.1.2248 (revision: 54483)

**How to reproduce**
```
create table my_prm_table (calc_year String not null) Engine = MergeTree() order by (calc_year);
insert into my_prm_table (calc_year) values ('2023');
insert into my_prm_table (calc_year) values ('2024');

create table my_values_table (id integer not null, calc_year String not null, amount Nullable(Float) default 0.)
Engine = MergeTree() order by (id) partition by (calc_year);

insert into my_values_table (id, calc_year, amount) values (1,'2023',10);
insert into my_values_table (id, calc_year, amount) values (2,'2023',20);
insert into my_values_table (id, calc_year, amount) values (3,'2024',30);
insert into my_values_table (id, calc_year, amount) values (4,'2024',40);


create or replace view v_my_test_view as
with cte_params as 
(select calc_year from my_prm_table
  where calc_year = {p_calc_year:String}
)
select * from my_values_table t
where t.calc_year in (select p.calc_year from cte_params p);

create or replace view v_my_test_view2 as
select * from my_values_table t
where t.calc_year in (select calc_year from my_prm_table p where p.calc_year = {p_calc_year:String});

create or replace view v_my_test_view3 as
select * from my_values_table t
where t.calc_year = {p_calc_year:String};

select * from v_my_test_view(p_calc_year='2024');
select * from v_my_test_view2(p_calc_year='2023');
select * from v_my_test_view3(p_calc_year='2023');

```

clickhouse restart

in logs error

```
2024.03.13 16:31:54.121523 [ 251590 ] {} <Error> Application: Code: 695. DB::Exception: Load job 'load table dev.v_my_test_view' failed: Code: 49. DB::Exception: Invalid storage definition in metadata file: it's a bug or result of manual intervention in metadata files: Cannot attach table `dev`.`v_my_test_view` from metadata file /var/lib/clickhouse/store/8a5/8a57e63c-de34-4b55-80d9-b6202241deb1/v_my_test_view.sql from query ATTACH VIEW dev.v_my_test_view UUID '5a2fe605-3587-4db7-86be-38e0c9080d57' AS WITH cte_params AS (SELECT calc_year FROM dev.my_prm_table WHERE calc_year = {p_calc_year:String}) SELECT * FROM dev.my_values_table AS t WHERE t.calc_year IN (SELECT p.calc_year FROM cte_params AS p). (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cf5565b
1. DB::Exception::Exception<>(int, FormatStringHelperImpl<>) @ 0x00000000078c7f43
2. DB::createTableFromAST(DB::ASTCreateQuery, String const&, String const&, std::shared_ptr<DB::Context>, DB::LoadingStrictnessLevel) @ 0x0000000010e41db4
3. DB::DatabaseOrdinary::loadTableFromMetadata(std::shared_ptr<DB::Context>, String const&, DB::QualifiedTableName const&, std::shared_ptr<DB::IAST> const&, DB::LoadingStrictnessLevel) @ 0x0000000010e63734
4. void std::__function::__policy_invoker<void (DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&)>::__call_impl<std::__function::__default_alloc_func<DB::DatabaseOrdinary::loadTableFromMetadataAsync(DB::AsyncLoader&, std::unordered_set<std::shared_ptr<DB::LoadJob>, std::hash<std::shared_ptr<DB::LoadJob>>, std::equal_to<std::shared_ptr<DB::LoadJob>>, std::allocator<std::shared_ptr<DB::LoadJob>>>, std::shared_ptr<DB::Context>, String const&, DB::QualifiedTableName const&, std::shared_ptr<DB::IAST> const&, DB::LoadingStrictnessLevel)::$_0, void (DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&)>>(std::__function::__policy_storage const*, DB::AsyncLoader&, std::shared_ptr<DB::LoadJob> const&) @ 0x0000000010e6e793
5. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::AsyncLoader::spawn(DB::AsyncLoader::Pool&, std::unique_lock<std::mutex>&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000d127fa6
6. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000cff84e1
7. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000cffbd1a
8. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000cffab1e
9. ? @ 0x00007fa7ccc06609
10. ? @ 0x00007fa7ccb2b353
 (version 24.2.1.2248 (official build)). (ASYNC_LOAD_FAILED), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cf5565b
1. DB::Exception::Exception<String const&, String>(int, FormatStringHelperImpl<std::type_identity<String const&>::type, std::type_identity<String>::type>, String const&, String&&) @ 0x00000000082b74fd
2. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::AsyncLoader::spawn(DB::AsyncLoader::Pool&, std::unique_lock<std::mutex>&)::$_0, void ()>>(std::__function::__policy_storage const*) @ 0x000000000d128409
3. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false>, void*>) @ 0x000000000cff84e1
4. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000cffbd1a
5. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000cffab1e
6. ? @ 0x00007fa7ccc06609
7. ? @ 0x00007fa7ccb2b353
 (version 24.2.1.2248 (official build))
2024.03.13 16:31:54.121602 [ 251590 ] {} <Information> Application: shutting down
2024.03.13 16:31:54.121608 [ 251590 ] {} <Debug> Application: Uninitializing subsystem: Logging Subsystem
2024.03.13 16:31:54.122262 [ 251591 ] {} <Trace> BaseDaemon: Received signal -2
2024.03.13 16:31:54.122308 [ 251591 ] {} <Information> BaseDaemon: Stop SignalListener thread
2024.03.13 16:31:54.166267 [ 251589 ] {} <Information> Application: Child process exited normally with code 183.

```

**Comment by kssenii** [comment]
cc @SmitaRKulkarni

**Comment by tavplubix** [comment]
Fixed by https://github.com/ClickHouse/ClickHouse/pull/60708

**Comment by jeanatpi** [comment]
Encountered the same issue. 
Is there a way to "fix" the issue when we have this error?

Deleting the metadata file ? And re-create the view ?

**Comment by areshayevcaci** [comment]
_>Deleting the metadata file ? And re-create the view_ 
this is what I'm doing - delete metadata, recreate view. if I need restart ch - do this again
the issue was reported as resolved and I hope it's fixed in the nearest build

**Comment by areshayevcaci** [comment]
checked - it was fixed in 24.3.2.23
